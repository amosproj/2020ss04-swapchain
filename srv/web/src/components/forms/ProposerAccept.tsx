
import React, { useState, useEffect } from "react"
import { fakeAddress } from "../../util"
import { PriorityChoice } from "./PriorityChoice"
import { Label } from "./Label"
import { Input } from "./Input"
import { SubmitButton } from "./SubmitButton"
import { getSecret } from "../../pkg/secret/secret"

export const ProposerAccept = () => {

    enum Priority {
        HIGH,
        MEDIUM,
        LOW,
    }

    const [fields, setFields] = useState({
        bitcoinPublicAddress: "",
        bitcoinPrivateKey: "",
        bitsharesAccountName: "",
        bitsharesPrivateKey: "",
        priority: Priority.HIGH,
        secret: getSecret(),
    })
    const [isValid, setValid] = useState(false)
    const [showError, setShowError] = useState(false)
    const [accepted, setAccepted] = useState(false)

    //validation
    useEffect(() => {
        const isValid = fields.bitcoinPublicAddress !== "" &&
            fields.bitcoinPrivateKey !== "" &&
            fields.bitsharesAccountName !== "" &&
            fields.bitsharesPrivateKey !== ""

        setValid(isValid)
        if (isValid) {
            setShowError(false)
        }
    }, [fields])

    const updateField = (e: React.FormEvent<HTMLInputElement>) => {
        setFields({
            ...fields,
            [e.currentTarget.name]: e.currentTarget.value
        })
    }
    const updateFieldByName = (name: string, value: number | string) => {
        setFields({
            ...fields,
            [name]: value
        })
    }

    const submitHandler = () => {
        if (!isValid) {
            setShowError(true)
            return
        }
        //TODO: @webersni

        setAccepted(true)
        console.log(fields)
    }

    return (
        <div className="container mx-auto">
            {accepted ?
                <div>

                    <section className="px-3 py-6">
                        <Label label="Your secret hash"></Label>
                        <p className="p-3 font-mono text-center text-teal-900 break-all border border-teal-400 rounded text-bold">{fields.secret.hash}</p>
                    </section>
                    <footer className="pt-12 mt-8 text-center border-t border-gray-300">
                        <p className="mt-3 text-sm text-gray-700">Copy this hash and give it to your trading partner</p>
                    </footer>
                </div>
                :
                <div>

                    <section className="mt-12 -mx-3 md:flex">
                        <div className="px-3 md:w-1/2">
                            <Label label="Bitcoin public address"></Label>
                            <Input name="bitcoinPublicAddress" value={fields.bitcoinPublicAddress} onChange={updateField} type="text" placeholder={fakeAddress(30)}></Input>
                        </div>
                        <div className="px-3 mt-4 md:w-1/2 md:mt-0">
                            <Label label="Bitshares account name"></Label>
                            <Input name="bitsharesAccountName" value={fields.bitsharesAccountName} onChange={updateField} type="text" placeholder="amos"></Input>
                        </div>
                    </section>
                    <section className="mt-12 ">
                        <div className="-mx-3 md:flex">
                            <div className="px-3 md:w-1/2">
                                <Label label="Bitcoin private key"></Label>
                                <Input name="bitcoinPrivateKey" value={fields.bitcoinPrivateKey} onChange={updateField} type="text" placeholder={fakeAddress(30)}></Input>
                            </div>
                            <div className="px-3 mt-4 md:w-1/2 md:mt-0">
                                <Label label="Bitshares private key"></Label>
                                <Input name="bitsharesPrivateKey" value={fields.bitsharesPrivateKey} onChange={updateField} type="text" placeholder={fakeAddress(30)}></Input>
                            </div>
                        </div>
                        <p className="mt-2 text-sm text-center text-gray-500 ">
                            Your private keys will never leave your browser, it is only used to sign your transactions. <a href="" className="relative text-xs text-blue-500">Read more in our docs.</a>
                        </p>
                    </section>
                    <section className="mt-12">
                        <Label label="Choose your priority"></Label>
                        <div className="flex flex-col items-center -mx-3 space-y-4 md:space-y-0 md:flex-row">
                            <button className="w-full px-3 md:w-1/3 focus:outline-none" onClick={() => updateFieldByName("priority", Priority.HIGH)}>
                                <PriorityChoice label={Priority[Priority.HIGH]} description="You pay the highest fees to increase the chance for your transaction to be picked up by the miners." selected={fields.priority === Priority.HIGH}></PriorityChoice>
                            </button>
                            <button className="w-full px-3 md:w-1/3 focus:outline-none" onClick={() => updateFieldByName("priority", Priority.MEDIUM)}>
                                <PriorityChoice label={Priority[Priority.MEDIUM]} description="You pay a moderate amount of fees so miners will probably confirm your transaction soon." selected={fields.priority === Priority.MEDIUM}></PriorityChoice>
                            </button>
                            <button className="w-full px-3 md:w-1/3 focus:outline-none" onClick={() => updateFieldByName("priority", Priority.LOW)}>
                                <PriorityChoice label={Priority[Priority.LOW]} description="You pay the lowest fees but might have to wait a few more blocks for your transaction to be confirmed." selected={fields.priority === Priority.LOW}></PriorityChoice>
                            </button>
                        </div>
                    </section>
                    <footer className="pt-12 mt-8 text-center border-t border-gray-300">
                        < p className={`-mt-4 pb-4 text-red-500 text-sm ${showError ? "" : "hidden"}`}>Please fill in all fields</p>
                        <SubmitButton color={isValid ? "teal" : "gray"} label="Submit" onClick={submitHandler}></SubmitButton>
                    </footer>
                </div>
            }
        </div >
    )
}