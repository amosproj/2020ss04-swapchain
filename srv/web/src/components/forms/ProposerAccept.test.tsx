import React from 'react';
import {ProposerAccept} from './ProposerAccept';
import renderer from 'react-test-renderer';

it('renders correctly', () => {
  const tree = renderer
    .create(<ProposerAccept></ProposerAccept>)
    .toJSON();
  expect(tree).toMatchSnapshot();
});