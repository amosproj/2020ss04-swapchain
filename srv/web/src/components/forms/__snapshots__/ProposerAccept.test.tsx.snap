// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="container mx-auto"
>
  <div>
    <section
      className="mt-12 -mx-3 md:flex"
    >
      <div
        className="px-3 md:w-1/2"
      >
        <label
          className="text-xs font-medium leading-4 tracking-wider text-gray-700 uppercase"
        >
          Bitcoin public address
        </label>
        <input
          className="w-full py-3 font-mono text-center text-gray-700 bg-gray-100 border border-gray-200 rounded focus:bg-white focus:border-teal-400"
          name="bitcoinPublicAddress"
          onChange={[Function]}
          placeholder="26b2309210010d82fdf72989ba67fd..."
          required={true}
          type="text"
          value=""
        />
      </div>
      <div
        className="px-3 mt-4 md:w-1/2 md:mt-0"
      >
        <label
          className="text-xs font-medium leading-4 tracking-wider text-gray-700 uppercase"
        >
          Bitshares account name
        </label>
        <input
          className="w-full py-3 font-mono text-center text-gray-700 bg-gray-100 border border-gray-200 rounded focus:bg-white focus:border-teal-400"
          name="bitsharesAccountName"
          onChange={[Function]}
          placeholder="amos"
          required={true}
          type="text"
          value=""
        />
      </div>
    </section>
    <section
      className="mt-12 "
    >
      <div
        className="-mx-3 md:flex"
      >
        <div
          className="px-3 md:w-1/2"
        >
          <label
            className="text-xs font-medium leading-4 tracking-wider text-gray-700 uppercase"
          >
            Bitcoin private key
          </label>
          <input
            className="w-full py-3 font-mono text-center text-gray-700 bg-gray-100 border border-gray-200 rounded focus:bg-white focus:border-teal-400"
            name="bitcoinPrivateKey"
            onChange={[Function]}
            placeholder="c57308699e78186a5a3dff412b0b2e..."
            required={true}
            type="text"
            value=""
          />
        </div>
        <div
          className="px-3 mt-4 md:w-1/2 md:mt-0"
        >
          <label
            className="text-xs font-medium leading-4 tracking-wider text-gray-700 uppercase"
          >
            Bitshares private key
          </label>
          <input
            className="w-full py-3 font-mono text-center text-gray-700 bg-gray-100 border border-gray-200 rounded focus:bg-white focus:border-teal-400"
            name="bitsharesPrivateKey"
            onChange={[Function]}
            placeholder="e82a689cc24029fc7703c7ff1d8c5d..."
            required={true}
            type="text"
            value=""
          />
        </div>
      </div>
      <p
        className="mt-2 text-sm text-center text-gray-500 "
      >
        Your private keys will never leave your browser, it is only used to sign your transactions. 
        <a
          className="relative text-xs text-blue-500"
          href="/"
        >
          Read more in our docs.
        </a>
      </p>
    </section>
    <section
      className="mt-12"
    >
      <label
        className="text-xs font-medium leading-4 tracking-wider text-gray-700 uppercase"
      >
        Choose your priority
      </label>
      <div
        className="flex flex-col items-center -mx-3 space-y-4 md:space-y-0 md:flex-row"
      >
        <button
          className="w-full px-3 md:w-1/3 focus:outline-none"
          onClick={[Function]}
        >
          <div
            className="relative flex justify-between items-center border rounded border-teal-400 shadow"
          >
            <div
              className="flex flex-col items-start p-4"
            >
              <span
                className="text-left font-bold text-gray-900"
              />
              <p
                className="text-sm text-left text-gray-600"
              >
                You pay the highest fees to increase the chance for your transaction to be picked up by the miners.
              </p>
            </div>
            <div
              className="flex flex-col items-center justify-center"
            >
              <span
                className="p-1 m-1 text-xs font-thin text-gray-500"
              />
              <span
                className="mb-4 mr-4 text-3xl font-bold hidden md:block text-gray-900 "
              >
                HIGH
              </span>
            </div>
          </div>
        </button>
        <button
          className="w-full px-3 md:w-1/3 focus:outline-none"
          onClick={[Function]}
        >
          <div
            className="relative flex justify-between items-center border rounded border-gray-200 bg-gray-100"
          >
            <div
              className="flex flex-col items-start p-4"
            >
              <span
                className="text-left font-bold text-gray-600"
              />
              <p
                className="text-sm text-left text-gray-400"
              >
                You pay a moderate amount of fees so miners will probably confirm your transaction soon.
              </p>
            </div>
            <div
              className="flex flex-col items-center justify-center"
            >
              <span
                className="p-1 m-1 text-xs font-thin text-gray-500"
              />
              <span
                className="mb-4 mr-4 text-3xl font-bold hidden md:block text-gray-300"
              >
                MEDIUM
              </span>
            </div>
          </div>
        </button>
        <button
          className="w-full px-3 md:w-1/3 focus:outline-none"
          onClick={[Function]}
        >
          <div
            className="relative flex justify-between items-center border rounded border-gray-200 bg-gray-100"
          >
            <div
              className="flex flex-col items-start p-4"
            >
              <span
                className="text-left font-bold text-gray-600"
              />
              <p
                className="text-sm text-left text-gray-400"
              >
                You pay the lowest fees but might have to wait a few more blocks for your transaction to be confirmed.
              </p>
            </div>
            <div
              className="flex flex-col items-center justify-center"
            >
              <span
                className="p-1 m-1 text-xs font-thin text-gray-500"
              />
              <span
                className="mb-4 mr-4 text-3xl font-bold hidden md:block text-gray-300"
              >
                LOW
              </span>
            </div>
          </div>
        </button>
      </div>
    </section>
    <footer
      className="pt-12 mt-8 text-center border-t border-gray-300"
    >
      <p
        className="-mt-4 pb-4 text-red-500 text-sm hidden"
      >
        Please fill in all fields
      </p>
      <button
        className="bg-gray-100 border-gray-400 inline-flex px-8 py-3 font-semibold text-gray-900 uppercase transition duration-200 ease-in-out border focus:outline-none rounded-sm hover:bg-white"
        onClick={[Function]}
        type="button"
      >
        Submit
      </button>
    </footer>
  </div>
</div>
`;
